pkg/minikube/cni/cni.go

```diff
@@ -109,13 +109,23 @@ func chooseDefault(cc config.ClusterConfig) Manager {
 		return Bridge{cc: cc}
 	}
 
-	if len(cc.Nodes) > 1 {
-		klog.Infof("%d nodes found, recommending kindnet", len(cc.Nodes))
-		return KindNet{cc: cc}
-	}
-
-	klog.Infof("CNI unnecessary in this configuration, recommending no CNI")
-	return Disabled{}
+	// Recommending KindNet for single and multi node clusters as well. This comes at a very small
+	// cost of having one extra KindNet pod running in all control plane node, but provides the
+	// following benefits:
+	//   - With Disabled CNI, Single node cluster doesn't respect the podCIDR even users ask for it
+	// via --extra-config=kubeadm.pod-network-cidr=10.244.0.0/16. This results in pods having ips
+	// like 172.17.0.2.
+	//
+	//   - Enables KindNet CNI in master in multi node cluster, This solves the network problem
+	// inside pod for multi node clusters. See https://github.com/kubernetes/minikube/issues/9838.
+	//
+	//   - Reduces the complexities for reconfiguring the control plane node, in a scenario
+	// where minikube cluster is created as a single node cluster and more node was added later.
+	//
+	// If a --cni flag is explicitly specified we do not do a node length check, this behaviour should
+	// also be persevered for chooseDefault.
+	klog.Infof("Recommending kindnet as default CNI", len(cc.Nodes))
+	return KindNet{cc: cc}
```

Comment:
for this PR could we keep the existing behaviour (single node have no CNI pod) ? 
a lot of our users will never need multinode

we could automaticalyy enable the CNi only if the user specifies the pod subnet through flags...otherwise for a lot of our IDE users I think having least amount of CPU usage and least amount of pods running be idea.

Reply:
Is there any known way to measure cpu and memory usages for minikube? I would like to compare the performances for these with and without kindnet.
What I am worried about is - if a minikube cluster is created with only single node now and more node was added later? how will we fix the cni in already running control plane nodes?

Reply:
thats a good quesiton @priyawadhwa has done a lot of work in this area. @tstromberg also has script for it.
It would be great if we had this on website or as a shell script to run manually .

Reply:
https://github.com/tstromberg/time-to-k8s

Reply:
https://github.com/tstromberg/time-to-k8s/blob/master/minikube-cni.yaml has a CNI benchmark test example you can use:

`go run . --config minikube-cni.yaml --iterations=4` should be sufficient to get the test results.

It outputs a CSV, which can be used for further analysis, such as: https://docs.google.com/spreadsheets/d/1oZq04X0QmHvucCLKNmLsvTJBdv75dOZnPP9dssOLteU/edit?usp=sharing

Reply:
Thanks for sharing those, let me take a look at them.

Reply:
https://docs.google.com/spreadsheets/d/1Rv7KCtVUYeMhFFzgY72J2T-ti3KDEdStggy3KyRfgt0/edit#gid=1614668143 has an initial graph, but it only has 2 iterations, so it shouldn't be considered authoritative.

Reply:
Thank you for the doc @tstromberg.

@medyagh while I am debugging why there is the extra delay setting up CNIs, how about an alternate solution to solve problem with multiple nodes. 

We set and store a cluster config something like `CreatedWithMultiNode bool`  true whenever user provides `--node > 1`. And we use this value to decide the CNI between KindNet and Disabled. And when a cluster is created with only control plane node and a node is added later, we show a warning that multi node networking might be broken (if it is not reconfigured already).